### PR TITLE
kernel: Tweak k_pipe_put() API to use 'const' with data parameter

### LINKED ISSUE #53946: `k_pipe_put`: data parameter should be const?
**Is your enhancement proposal related to a problem? Please describe.**
Docs: https://docs.zephyrproject.org/3.1.0/kernel/services/data_passing/pipes.html#c.k_pipe_put
The data parameter not being marked as const limits the data sources that can be passed to pipes. I use const where I can and the lack of const here makes pipes very difficult to use correctly

The [message queue equivalent is const](https://docs.zephyrproject.org/3.1.0/kernel/services/data_passing/message_queues.html#c.k_msgq_put). It would be highly irregular to actually write to the source buffer during the put operation, and adding const has no backward compatibility issues (it only increases the valid data sources)

### PR DESCRIPTION
**Pull request by peter-mitsis**
In the function _k_pipe_put()_, the _data_ parameter, which points to a buffer that is written to the pipe, now uses the `const` keyword. Internally, that pointer is cast to drop the 'const' to work with the common code that both _k_pipe_put()_ and _k_pipe_get()_ share as it is known that k_pipe_put() will take a path where the contents of that buffer will not be modified.

Fixes #53946

### CODE REVIEW

**Review by andyross — DISMISSED** [review]
Makes sense.  To short circuit any pedantry: adding a const to a function pointer argument is never a breaking API change except to code within that function.  Existing code passing non-const pointers will be fine, they'll promote automatically.

File: tests/kernel/pipe/pipe_api/src/test_pipe_contexts.c
Comment (peter-mitsis) [reply tests/kernel/pipe/pipe_api/src/test_pipe_contexts.c]: @andyross - Comment removed. (Unfortunately that dismissed the review).

**Review by cfriedt — APPROVED** [review]

**Review by andyross — APPROVED** [review]
refresh +1

Also repeat the standard note: I don't really like our auto-dismiss-reviews-on-push policy and am happy if those with merge rights treat a previous +1 of mine as live.  I trust people to ping for genuine re-review on big changes.  Also I'm an inveterate flake and if we make all these patches wait on my senseless re-review we'll only slow the project down.

**Review by nashif — APPROVED** [review]

**Pull request merged**
